worker/caasapplicationprovisioner/application.go

```diff
@@ -97,20 +96,93 @@ func (a *appWorker) Wait() error {
 }
 
 func (a *appWorker) loop() error {
-	charmURL, err := a.facade.ApplicationCharmURL(a.name)
-	if errors.IsNotFound(err) {
-		a.logger.Debugf("application %q removed", a.name)
-		return nil
-	} else if err != nil {
-		return errors.Annotatef(err, "failed to get charm url for application")
-	}
-	charmInfo, err := a.facade.CharmInfo(charmURL.String())
+	// Wait till charm is upgraded to a v2 charm.
+	appStateWatcher, err := a.facade.WatchApplication(a.name)
 	if err != nil {
-		return errors.Annotatef(err, "failed to get application charm deployment metadata for %q", a.name)
+		return errors.Annotatef(err, "failed to watch for changes to application %q", a.name)
 	}
-	if charmInfo == nil ||
-		corecharm.Format(charmInfo.Charm()) < corecharm.FormatV2 {
-		return errors.Errorf("charm version 2 or greater required")
+	if err := a.catacomb.Add(appStateWatcher); err != nil {
```

Comment:
Let's pluck this out into a function called something like verifyCharmUpgraded. This function is fine being large for key code paths relating to the main select statement, but this is something that can be isolated.

Reply:
Good call, done.